tests/datamodules/test_experience_sources.py

```diff
@@ -32,123 +39,285 @@ def test_iterator(self):
-        self.xp_stream.episodes = 1
-
-        for i_batch, batch in enumerate(self.rl_dataloader):
-            self.assertEqual(len(batch), 1)
-            self.assertIsInstance(batch[0][0], Experience)
-            self.assertEqual(batch[0][-1].done, True)
-
-    def test_experience_stream_MULTI_EPISODE(self):
-        """Check that the experience stream gives 4 full episodes per batch"""
-        self.xp_stream.episodes = 4
-
-        for i_batch, batch in enumerate(self.rl_dataloader):
-            self.assertEqual(len(batch), 4)
-            self.assertIsInstance(batch[0][0], Experience)
-            self.assertEqual(batch[0][-1].done, True)
+        self.agent = DummyAgent(net=self.net)
+        self.env = [gym.make("CartPole-v0") for _ in range(2)]
+        self.device = torch.device('cpu')
+        self.n_steps = 3
+        self.gamma = 0.9
+        self.source = DiscountedExperienceSource(self.env, self.agent, n_steps=self.n_steps, gamma=self.gamma)
+
+        self.s1 = torch.ones(3)
+        self.s2 = torch.zeros(3)
+
+        self.exp1 = Experience(state=self.s1, action=1, reward=1, done=False, new_state=self.s2)
+        self.exp2 = Experience(state=self.s2, action=1, reward=1, done=False, new_state=self.s1)
+
+        self.env1 = Mock()
+        self.env1.step = Mock(return_value=(self.s2, 1, True, self.s1))
+
+    def test_init(self):
+        """Test that experience source is setup correctly"""
+        self.assertEqual(self.source.n_steps, self.n_steps + 1)
+        self.assertEqual(self.source.steps, self.n_steps)
+        self.assertEqual(self.source.gamma, self.gamma)
+
+    def test_source_step(self):
+        """Tests that the source returns a single experience"""
+
+        for idx, exp in enumerate(self.source.stepper(self.device)):
+            self.assertTrue(isinstance(exp, Experience))
+            break
+
+    def test_source_step_done(self):
+        """Tests that the source returns a single experience"""
+
+        self.device = torch.device('cpu')
+        self.source = DiscountedExperienceSource(self.env1, self.agent, n_steps=self.n_steps)
+
+        self.source.histories[0].append(self.exp1)
+        self.source.histories[0].append(self.exp2)
+
+        for idx, exp in enumerate(self.source.stepper(self.device)):
+            self.assertTrue(isinstance(exp, Experience))
+            self.assertTrue(torch.all(torch.eq(exp.new_state, self.s2)))
+            break
+
+    def test_source_discounted_return(self):
+        """Tests that the source returns a single experience"""
+
+        self.device = torch.device('cpu')
+        self.source = DiscountedExperienceSource(self.env1, self.agent, n_steps=self.n_steps)
+
+        self.source.histories[0].append(self.exp1)
+        self.source.histories[0].append(self.exp2)
+
+        for idx, exp in enumerate(self.source.stepper(self.device)):
+            self.assertTrue(isinstance(exp, Experience))
+            self.assertEqual(exp.reward, 2.9701)
```

Comment:
mind add a comment on how the number comes up?

Reply:
will do